extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -0,0 +1,706 @@
+   * @private
+   */
+  maybeRenderAmpAd_(valid, bytes) {
+    if (!valid) {
+      return Promise.resolve(false);
+    }
+    // Timer id will be set if we have entered layoutCallback at which point
+    // 3p throttling count was incremented.  We want to "release" the throttle
+    // immediately since we now know we are not a 3p ad.
+    if (this.timerId_) {
+      decrementLoadingAds(this.timerId_, this.getWin());
+    }
+    return xhrFor(this.getWin()).utf8FromArrayBuffer(bytes).then(creative => {
+      // Find the json blob located at the end of the body and parse it.
+      const creativeMetaData = this.getAmpAdMetadata_(creative);
+      if (!creativeMetaData || !this.supportsShadowDom()) {
+        // Shadow DOM is not supported or could not find appropriate markers
+        // within the creative therefore load within cross domain iframe.
+        // Iframe is created immediately (as opposed to waiting for
+        // layoutCallback) as the the creative has been verified as AMP and
+        // will run efficiently.
+        this.renderViaIframe_();
+        this.rendered_ = true;
+        return Promise.resolve(true);
+      } else {
+        try {
+          // Do extraction processing on CSS and body before creating the
+          // shadow root so that if they error out, we don't actually edit
+          // the doc.
+          const cssBlock = this.formatCSSBlock_(creative, creativeMetaData);
+          const bodyBlock = this.formatBody_(creative, creativeMetaData);
+          // Copy fonts to host document head.
+          this.relocateFonts_(creativeMetaData);
+          // Add extensions to head.
+          this.addCreativeExtensions_(creativeMetaData);
+          // Finally, add body and re-formatted CSS styling to the shadow root.
+          const shadowRoot =
+              this.element.shadowRoot || this.element.createShadowRoot();
+          shadowRoot./*OK*/innerHTML += (cssBlock + bodyBlock);
+          this.rendered_ = true;
+          this.onAmpCreativeShadowDomRender();
+          return Promise.resolve(true);
+        } catch (e) {
+          // If we fail on any of the steps of Shadow DOM construction, just
+          // render in iframe.
+          // TODO: report!
+          return Promise.resolve(false);
+        }
+      }
+    });
+  }
+
+  /**
+   * Creates iframe whose src matches that of the ad URL.  The response should
+   * have been cached causing the browser to render without callout.  However,
+   * it is possible for cache miss to occur which can be detected server-side
+   * by missing ORIGIN header.
+   * @param {boolean=} opt_isNonAmpCreative whether creative within iframe
+   *    is AMP creative (if not, intersection observer allows sending info into
+   *    nested frames).
+   * @private
+   */
+  renderViaIframe_(opt_isNonAmpCreative) {
+    if (!this.adUrl_) {
+      // Error should not occur.
+      return;
+    }
+    const iframe = this.element.ownerDocument.createElement('iframe');
```

Comment:
Ok, @keithwrightbos and I refactored the handling of the AMP Ads API so that we now support the iframe postMessage API in both 3p-impl and amp-a4a.  (Whew!)  Also, to address the scheduling question raised elsewhere, per offline conversation with @cramforce , we're continuing to issue the XHR and promise chain in onLayoutMeasure(), but we added vsync.mutate() protection to all DOM mutations to prevent user-visible redraws.

PTAL.